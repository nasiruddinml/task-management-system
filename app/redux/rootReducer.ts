import { commonReducer, commonSlice } from './slices/common';
import { taskReducer, taskSlice } from './slices/task';

const reducers = {
  [commonSlice.name]: commonReducer,
  [taskSlice.name]: taskReducer,
};

export default reducers;
