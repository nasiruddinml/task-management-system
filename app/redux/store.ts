import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { Reducer } from 'redux';
import {
  FLUSH,
  PAUSE,
  PERSIST,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist';
import { RESET_STATE_ACTION_TYPE } from './actions/resetState';
import reducers from './rootReducer';

const combinedReducer = combineReducers<typeof reducers>(reducers);

export const rootReducer: Reducer<RootState> = (state, action) => {
  if (action.type === RESET_STATE_ACTION_TYPE) {
    state = {} as RootState;
  }

  return combinedReducer(state, action);
};

export const store = configureStore({
  preloadedState: {},
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
});

export const persistor = persistStore(store);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof combinedReducer>;
export const useTypedDispatch = () => useDispatch<AppDispatch>();
export const useTypedSelector: TypedUseSelectorHook<RootState> = useSelector;
