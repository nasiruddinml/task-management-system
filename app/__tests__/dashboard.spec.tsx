import { render, screen, fireEvent } from '@testing-library/react';
import Dashboard from '@app/(route pages)/dashboard/page';
import { Providers } from '@app/components/molecules/Providers/Providers';
import { rootReducer, store } from '@app/redux/store';
import { resetStateAction } from '@app/redux/actions/resetState';

describe('Dashboard', () => {
  let confirmSpy: any;
  beforeAll(() => {
    confirmSpy = jest.spyOn(window, 'confirm');
    confirmSpy.mockImplementation(jest.fn(() => true));
  });
  afterAll(() => {
    confirmSpy.mockRestore();
  });
  beforeEach(() => {
    jest.clearAllMocks();
    store.dispatch(resetStateAction);
  });

  it('renders dashboard unchanged', () => {
    const { container } = render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    expect(container).toMatchSnapshot();
  });

  it('render app title', async () => {
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );

    const heading = screen.getByText('Task Management App');
    expect(heading).toBeInTheDocument();
  });

  it('render state cards', async () => {
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );

    const heading = await screen.findByText('Total Tasks');
    expect(heading).toBeInTheDocument();
    const heading2 = await screen.findByText('Total Days');
    expect(heading2).toBeInTheDocument();
    const heading3 = await screen.findByText('Total Hours');
    expect(heading3).toBeInTheDocument();
  });

  it('test task add', () => {
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: 'testing task title' } });
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    expect(screen.getByText('testing task title')).toBeInTheDocument();
    const deleteButton = screen.getByRole('button', { name: /Delete/i });
    fireEvent.click(deleteButton);
  });

  it('test delete action', async () => {
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: 'testing task title' } });
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    const deleteButton = screen.getByRole('button', { name: /Delete/i });
    fireEvent.click(deleteButton);
    expect(confirmSpy).toHaveBeenCalled();
    expect(screen.getByText('No todo list! Hooray!!')).toBeInTheDocument();
  });

  it('test delete action cancel confirm', async () => {
    confirmSpy.mockImplementation(jest.fn(() => false));
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: 'testing task title' } });
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    const deleteButton = screen.getByRole('button', { name: /Delete/i });
    fireEvent.click(deleteButton);
    expect(confirmSpy).toHaveBeenCalled();
    expect(screen.getByText('testing task title')).toBeInTheDocument();
    confirmSpy.mockImplementation(jest.fn(() => true));
    fireEvent.click(deleteButton);
  });

  it('test total task', async () => {
    confirmSpy.mockImplementation(jest.fn(() => false));
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: 'testing task title' } });
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    // eslint-disable-next-line testing-library/no-node-access
    expect(screen.getByText('Total Tasks').nextElementSibling?.innerHTML).toBe(
      '1',
    );
    const deleteButton = screen.getByRole('button', { name: /Delete/i });
    fireEvent.click(deleteButton);
  });

  it('test total days', async () => {
    confirmSpy.mockImplementation(jest.fn(() => false));
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: 'testing task title' } });
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    // eslint-disable-next-line testing-library/no-node-access
    expect(screen.getByText('Total Days').nextElementSibling?.innerHTML).toBe(
      '1',
    );
    const deleteButtons = screen.getAllByRole('button', { name: /Delete/i });
    deleteButtons.forEach(elm => fireEvent.click(elm));
  });

  it('test total hours', async () => {
    confirmSpy.mockImplementation(jest.fn(() => false));
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: 'testing task title' } });
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    // eslint-disable-next-line testing-library/no-node-access
    expect(screen.getByText('Total Hours').nextElementSibling?.innerHTML).toBe(
      '12',
    );
    fireEvent.change(titleInput, { target: { value: 'testing task title' } });
    fireEvent.change(timeInput, { target: { value: 4 } });
    fireEvent.click(addButton);
    // eslint-disable-next-line testing-library/no-node-access
    expect(screen.getByText('Total Hours').nextElementSibling?.innerHTML).toBe(
      '16',
    );
    const deleteButtons = screen.getAllByRole('button', { name: /Delete/i });
    deleteButtons.forEach(elm => fireEvent.click(elm));
  });

  it('test input error for title empty', async () => {
    confirmSpy.mockImplementation(jest.fn(() => false));
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: '' } });
    fireEvent.blur(titleInput);
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    expect(screen.getByText('Task title is required')).toBeInTheDocument;
    const deleteButtons = screen.getAllByRole('button', { name: /Delete/i });
    deleteButtons.forEach(elm => fireEvent.click(elm));
  });

  it('test input error for long text', async () => {
    confirmSpy.mockImplementation(jest.fn(() => false));
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, {
      target: {
        value:
          "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.",
      },
    });
    fireEvent.blur(titleInput);
    fireEvent.change(timeInput, { target: { value: 4 } });
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    expect(screen.getByText('Task title must have a maximum of 128 characters'))
      .toBeInTheDocument;
    const deleteButtons = screen.getAllByRole('button', { name: /Delete/i });
    deleteButtons.forEach(elm => fireEvent.click(elm));
  });

  it('test input error for time ', async () => {
    confirmSpy.mockImplementation(jest.fn(() => false));
    render(
      <Providers>
        <Dashboard />
      </Providers>,
    );
    const titleInput = screen.getByLabelText('Title field');
    const timeInput = screen.getByLabelText('time field');
    fireEvent.change(titleInput, { target: { value: 'Lorem Ipsum' } });
    fireEvent.change(timeInput, { target: { value: -1 } });
    fireEvent.blur(timeInput);
    const addButton = screen.getByRole('button', { name: /Add/i });
    fireEvent.click(addButton);
    expect(screen.getByText('Time Required must be in between 0 to 24'))
      .toBeInTheDocument;
    const deleteButtons = screen.getAllByRole('button', { name: /Delete/i });
    deleteButtons.forEach(elm => fireEvent.click(elm));
  });
});
