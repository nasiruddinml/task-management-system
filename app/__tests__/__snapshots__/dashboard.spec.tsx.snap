// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders dashboard unchanged 1`] = `
<div>
  <header
    class="header"
  >
    <h2
      class="title"
    >
      Task Management App
    </h2>
    <div
      class="OverviewSection"
    >
      <div
        class="Card"
      >
        <p
          class="CardTitle"
        >
          Total Tasks
        </p>
        <span
          class="CardNumber"
        >
          0
        </span>
      </div>
      <div
        class="Card"
      >
        <p
          class="CardTitle"
        >
          Total Days
        </p>
        <span
          class="CardNumber"
        >
          0
        </span>
      </div>
      <div
        class="Card"
      >
        <p
          class="CardTitle"
        >
          Total Hours
        </p>
        <span
          class="CardNumber"
        >
          0
        </span>
      </div>
    </div>
  </header>
  <div
    class="content"
  >
    <form
      class="form"
    >
      <div
        class="formGroup"
      >
        <label
          class="formLabel"
        >
          Task Title
        </label>
        <input
          aria-label="Title field"
          class="formField"
          name="title"
          type="text"
          value=""
        />
      </div>
      <div
        class="formGroup hell"
      >
        <label
          class="formLabel"
        >
          Time Required (In Hrs)
        </label>
        <input
          aria-label="time field"
          class="formField"
          name="time"
          type="number"
          value="0"
        />
      </div>
      <div
        class="formActions"
      >
        <button
          class="button formSubmitBtn"
          type="submit"
        >
          Add
        </button>
      </div>
    </form>
    <div
      class="Card"
    >
      <h3>
        Todo list
      </h3>
      <div
        class="TableWrapper"
      >
        <p>
          No todo list! Hooray!!
        </p>
      </div>
    </div>
  </div>
</div>
`;
